src/views/open-api-docs-view-v2/server.ts

```diff
@@ -3,138 +3,314 @@
+import type { GetStaticPaths } from 'next'
+import type { GithubDir } from 'lib/fetch-github-file-tree'
 import type {
+	ApiDocsUrlContext,
+	OpenApiDocsV2Params,
+	OpenApiDocsViewV2Config,
 	OpenApiDocsViewV2Props,
 	SharedProps,
-	OpenApiDocsViewV2Config,
 } from 'views/open-api-docs-view-v2/types'
-import { OpenAPIV3 } from 'openapi-types'
+import type { OpenAPIV3 } from 'openapi-types'
+
+/**
+ * Generate static paths for an OpenAPI docs view.
+ */
+export async function generateStaticPaths({
+	schemaSource,
+	schemaTransforms = [],
+	transformVersionData = (versionData) => versionData,
+}: {
+	schemaSource: ApiDocsVersionData[] | GithubDir
+	schemaTransforms?: ((schema: OpenAPIV3.Document) => OpenAPIV3.Document)[]
+	transformVersionData?: (
+		versionData: ApiDocsVersionData[]
+	) => ApiDocsVersionData[]
+}): Promise<ReturnType<GetStaticPaths<OpenApiDocsV2Params>>> {
+	// If we are in a product repo deploy preview, don't pre-render any paths
+	if (isDeployPreview()) {
+		return { paths: [], fallback: 'blocking' }
+	}
+	/**
+	 * If we're in production, statically render the non-versioned landing view,
+	 * as well as the non-versioned operation views.
+	 *
+	 * We use `fallback: blocking` for versioned views. We could in theory
+	 * statically render all pages across all versions, but this would increase
+	 * our build times.
+	 *
+	 * We fetch and parse the default version of the OpenAPI schema to figure
+	 * out which operation slugs to statically render. Note the "default version"
+	 * is the latest stable version, or if there are no stable versions,
+	 * then the latest version regardless of release stage.
+	 */
+	// Fetch version data
+	const rawVersionData = Array.isArray(schemaSource)
+		? schemaSource
+		: await fetchCloudApiVersionData(schemaSource)
+	const versionData = transformVersionData(rawVersionData)
+	// Determine the default version
+	const defaultVersion = findDefaultVersion(versionData)
+	// Parse the default version
+	const schemaFileString =
+		typeof defaultVersion.sourceFile === 'string'
+			? defaultVersion.sourceFile
+			: await fetchGithubFile(defaultVersion.sourceFile)
+	const schemaData = await parseAndValidateOpenApiSchema(
+		schemaFileString,
+		schemaTransforms
+	)
+	// Extract operation objects, and map to slugs
+	const operationObjects = getOperationObjects(schemaData)
+	const operationSlugs = operationObjects.map((operation) => {
+		return slugifyOperationId(operation.operationId)
+	})
+	// Generate path objects for each operation slug
+	const pathObjects = operationSlugs.map((operationSlug) => {
+		return { params: { page: [operationSlug] } }
+	})
+	// Return paths
+	return {
+		paths: [{ params: { page: [] } }, ...pathObjects],
+		fallback: 'blocking',
+	}
+}
+
+/**
+ * Wrapper around `generateStaticProps` that handles the common production
+ * use case of fetching version data from GitHub.
+ */
+export async function generateStaticPropsVersioned(
```

Comment:
Second, we introduce `generateStaticPropsVersioned`, which handles the oft-repeated use case of fetching versioned OpenAPI schemas from GitHub, as we do for all of our HCP API docs.